### PR TITLE
IGNITE-14567

### PR DESCRIPTION
**Pull request by ascherbakoff**

### CODE REVIEW

File: modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+        return false;
+    }
+
+    /** */
+    private static Condition toCondition(ConditionInfo info) {
+        byte[] key = info.key();
+
+        ConditionType type = info.type();
+
+        if (type == ConditionType.KEY_EXISTS)
```
Comment (AMashenkov) [inline modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:329]: `Switch(type)` construction will be more natural and more effective here

File: modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+    /**
+     * @param storage Storage.
+     */
+    public MetaStorageListener(KeyValueStorage storage) {
+        this.storage = storage;
+        this.cursors = new ConcurrentHashMap<>();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onRead(Iterator<CommandClosure<ReadCommand>> iter) {
```
Comment (sanpwc) [inline modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:88]: Both for onRead() and onWrite(). Seems that within current implementation we don't propagate exceptions to client side.
Say CompactedException was thrown on storage.get(). Within given implementation we will log it with
```
            catch (Exception e) {
                LOG.error("Failed to process the command", e);
            }
```
in RaftServerImpl#processQueue
On the client side we'll hang on metaStorageSvc.get(EXPECTED_RESULT_ENTRY.key()).get() until TimeoutException is thrown.
I believe that we still need catch block that will clo.result(businessLogicException) to user.

File: modules/network/src/main/java/org/apache/ignite/network/MessagingService.java [hunk 3]
```diff
@@ -61,12 +61,23 @@
...
+    /**
+     * Sends a message asynchronously with same guarantees as {@link #send(ClusterNode, NetworkMessage)} and
+     * returns a future that will be completed successfully upon receiving a response.
+     *
+     * @param addr Recipient network address in host:port format.
+     * @param msg A message.
+     * @param timeout Waiting for response timeout in milliseconds.
+     * @return A future holding the response or error if the expected response was not received.
+     */
+    CompletableFuture<NetworkMessage> invoke(String addr, NetworkMessage msg, long timeout);
```
Comment (sanpwc) [inline modules/network/src/main/java/org/apache/ignite/network/MessagingService.java:79]: Why we need it?

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/message/impl/ActionRequestImpl.java [hunk 4]
```diff
@@ -22,20 +22,27 @@
...
 class ActionRequestImpl<T> implements ActionRequest, ActionRequest.Builder {
+    /** */
+    private String groupId;
+
     /** */
     private Command cmd;
 
     /** */
-    private String groupId;
+    private boolean readOnlySafe;
```
Comment (sanpwc) [inline modules/raft-client/src/main/java/org/apache/ignite/raft/client/message/impl/ActionRequestImpl.java:32]: /** {@inheritDoc} */

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/message/ActionRequest.java [hunk 5]
```diff
@@ -35,19 +35,30 @@
      */
     Command command();
 
+    /**
+     * @return {@code True} for linearizable reading.
+     */
+    boolean readOnlySafe();
```
Comment (sanpwc) [inline modules/raft-client/src/main/java/org/apache/ignite/raft/client/message/ActionRequest.java:41]: Currently there are no usages of readOnlySafe(). Why we need it? In Which cases do you expect to use non-linearizable readings? Besides that it's a bit confusing cause readOnlySafe only has sense for read operations, but ActrionRequest both covers reads and writes. In other words within given abstaction it's possible to create write request with readOnlySafe set to false. Do we expect non-linearizable writes?

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/CommandClosure.java [hunk 6]
```diff
@@ -30,14 +32,8 @@
...
-    void success(Object res);
-
-    /**
-     * Failure outcome.
-     * @param err The error.
+     * Must be called after a command has been processed normally.
+     * @param res Execution result.
      */
-    void failure(Throwable err);
+    void result(Object res);
```
Comment (sanpwc) [inline modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/CommandClosure.java:38]: Should we make an intention that res should be marshalable more obvious?

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupListener.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+import org.apache.ignite.raft.client.WriteCommand;
+
+/**
+ * A listener for replication group events.
+ */
+public interface RaftGroupListener {
+    /**
+     * The callback to apply read commands.
+     * <p>
+     * If the runtime exception is thrown during iteration all unprocessed read requests will be aborted with the STM
```
Comment (sanpwc) [inline modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupListener.java:32]: I believe that is relates only to unhandled runtime exceptions, cause nothing is wrong in further processing reads and writes in case of business logic runtime exceptions that were properly handled and propagated to client side with the help of clo.result(). For example, as was mentioned above CompactionException on metaStorage.get() shouldn't prevent further processing but should be propagated to client service.

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java [hunk 8]
```diff
@@ -99,6 +100,19 @@ public void stopRaftGroup(String groupId, List<ClusterNode> peers) {
         //Now we are using only one node in a raft group.
         //TODO: IGNITE-13885 Investigate jraft implementation for replication framework based on RAFT protocol.
         if (peers.get(0).name().equals(clusterNetSvc.topologyService().localMember().name()))
-            raftServer.clearListener(groupId);
+            raftServer.stopRaftGroup(groupId);
+    }
+
+    public RaftGroupService startRaftService(String groupId, List<ClusterNode> peers) {
```
Comment (sanpwc) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java:106]: javadoc

File: modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java (hunk 1 above)
Comment (ascherbakoff) [reply modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:329]: Agreed, but this is actually not my code. 
Class was renamed by me. 
Let's fix it by another ticket.

File: modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java (hunk 2 above)
Comment (ascherbakoff) [reply modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:88]: This is true, but as soon as jraft will be merged I going to remove RaftServerImpl.
This can be fixed by adding internal error callback to CommandClosureEx, but I think it's waste of time to fix it.

File: modules/network/src/main/java/org/apache/ignite/network/MessagingService.java (hunk 3 above)
Comment (ascherbakoff) [reply modules/network/src/main/java/org/apache/ignite/network/MessagingService.java:79]: To clearly separate messaging from discovery.
You don't need a node to be discovered to send a message to it and it's better fits to raft communication model, code is much cleaner. Moreover, ClusterNode has the address [1]
[1] org.apache.ignite.network.ClusterNode#address

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/message/ActionRequest.java (hunk 5 above)
Comment (ascherbakoff) [reply modules/raft-client/src/main/java/org/apache/ignite/raft/client/message/ActionRequest.java:41]: Usages are in jraft branch.

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java [hunk 9]
```diff
@@ -88,6 +92,9 @@
     /** */
     private final ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
 
+    /** */
+    private final boolean reuse;
```
[the remark this answers is not in the document]
Comment (ascherbakoff) [reply modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java:96]: It's used for tests currently, see [1]
Helps to reduce boilerplate code.

[1] org.apache.ignite.raft.server.ITRaftCounterServerTest

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupListener.java (hunk 7 above)
Comment (ascherbakoff) [reply modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupListener.java:32]: It's true. Javadoc directly tells about runtime exceptions:
If the **runtime** exception ...

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java [hunk 10]
```diff
@@ -287,67 +324,124 @@ public RaftGroupServiceImpl(
...
+    /**
+     * Retries request until success or timeout.
+     *
+     * @param addr Target address.
+     * @param req The request.
+     * @param stopTime Stop time.
+     * @param fut The future.
+     * @param <R> Response type.
+     */
+    private <R> void sendWithRetry(Peer peer, Object req, long stopTime, CompletableFuture<R> fut) {
```
Comment (sashapolo) [inline modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java:362]: Why did you change the previous approach for implementing this method? I think it looked more elegant and readable. It is possible to implement the current logic in a similar way. Here's what I as able to come with:

```
private <R> CompletableFuture<R> sendWithRetry(Peer peer, Object req, long stopTime) {
    if (currentTimeMillis() >= stopTime) {
        return CompletableFuture.failedFuture(new TimeoutException());
    }

    return cluster.messagingService().invoke(peer.address(), (NetworkMessage)req, timeout)
        .handle((resp, err) -> {
            if (err != null) {
                if (recoverable(err)) {
                    return composeWithDelay(() -> this.<R>sendWithRetry(randomNode(), req, stopTime));
                }
                else {
                    return CompletableFuture.<R>failedFuture(err);
                }
            }
            else if (resp instanceof RaftErrorResponse) {
                RaftErrorResponse resp0 = (RaftErrorResponse)resp;
                RaftErrorCode errorCode = resp0.errorCode();

                if (errorCode == null) {
                    leader = peer;

                    return CompletableFuture.<R>completedFuture(null);
                }

                switch (errorCode) {
                    case NO_LEADER:
                        return composeWithDelay(() -> this.<R>sendWithRetry(randomNode(), req, stopTime));
                    case LEADER_CHANGED:
                        leader = resp0.newLeader(); // Update a leader.

                        return composeWithDelay(() -> this.<R>sendWithRetry(resp0.newLeader(), req, stopTime));
                    default:
                        return CompletableFuture.<R>failedFuture(new RaftException(errorCode));
                }
            }
            else {
                leader = peer;

                return CompletableFuture.completedFuture((R)resp);
            }
        })
        .thenCompose(Function.identity());
}
```

What do you think?

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/message/ActionRequest.java (hunk 5 above)
Comment (ascherbakoff) [reply modules/raft-client/src/main/java/org/apache/ignite/raft/client/message/ActionRequest.java:41]: This concept comes from RAFT linerializable reading guarantees, which require stable leader before reading. 
It makes no sense for single node RAFT cluster, so no getter usage here.
You can find actual usage in ignite-13885

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java (hunk 10 above)
Comment (ascherbakoff) [reply modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java:359]: Because it doesn't work as intended. I've added a test reproducing a problem: testUserRequestLeaderElectedAfterDelayWithFailedNode.

Actually I prefer to not change current approach, because: 1) it works for me 2) it looks quite readable to me 3) it's a bit more efficient (creates less amount of futures. 4) I think your solution is not perfect either and can be improved.

I'm ok to address this later in separate ticket, after merging jraft.

**Review by sashapolo — APPROVED** [review]
